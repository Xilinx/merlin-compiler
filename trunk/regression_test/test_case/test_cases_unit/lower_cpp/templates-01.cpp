//(C) Copyright 2016-2021 Xilinx, Inc.
//All Rights Reserved.
//
//Licensed to the Apache Software Foundation (ASF) under one
//or more contributor license agreements.  See the NOTICE file
//distributed with this work for additional information
//regarding copyright ownership.  The ASF licenses this file
//to you under the Apache License, Version 2.0 (the
//"License"); you may not use this file except in compliance
//with the License.  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing,
//software distributed under the License is distributed on an
//"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
//KIND, either express or implied.  See the License for the
//specific language governing permissions and limitations
//under the License. (edited)
extern void printf(const char *, ...);

template <typename T> T factorial(T n) {
    return n > 0 ? n * factorial(n - 1) : 1;
}

template <typename T, typename U> T factorial(T n, U u) {
    return n > 0 ? n * factorial(n - 1) * factorial(n - 1, u) : 1;
}

// should be ignored.
template <> unsigned char factorial(unsigned char);

#pragma ACCEL kernel
int main() {
    unsigned arg = 23;
    // implicit instantiation.
    printf("%d\n", factorial(arg));
    printf("%d\n", factorial(arg, true));
}
