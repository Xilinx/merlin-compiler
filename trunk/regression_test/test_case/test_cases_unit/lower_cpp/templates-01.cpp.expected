extern void printf(const char *,... );
unsigned char p_Z9factorialIhEh(unsigned char n);
unsigned int p_Z9factorialIjEj(unsigned int n);

unsigned int p_Z9factorialIjEj(unsigned int n)
{
  return n > ((unsigned int )0)?n * p_Z9factorialIjEj(n - ((unsigned int )1)) : ((unsigned int )1);
}
static void __merlin_dummy_SgTemplateFunctionDeclaration_template_();
unsigned int p_Z9factorialIjbEjb(unsigned int n,bool u);

unsigned int p_Z9factorialIjbEjb(unsigned int n,bool u)
{
  return n > ((unsigned int )0)?n * p_Z9factorialIjEj(n - ((unsigned int )1)) * p_Z9factorialIjbEjb(n - ((unsigned int )1),u) : ((unsigned int )1);
}
static void __merlin_dummy_SgTemplateFunctionDeclaration_template_();
// should be ignored.
static void __merlin_dummy_SgTemplateInstantiationFunctionDecl_unsigned_c_();
#pragma ACCEL kernel

int main()
{
  unsigned int arg = 23;
// implicit instantiation.
  printf("%d\n",(p_Z9factorialIjEj(arg)));
  printf("%d\n",(p_Z9factorialIjbEjb(arg,true)));
}
